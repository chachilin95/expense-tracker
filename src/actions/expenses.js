import database from '../firebase/firebase';

////////// ADD EXPENSE //////////

export const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
});

export const startAddExpense = (expenseData = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        const {
            description = '', 
            note = '', 
            amount = 0, 
            createdAt = 0 
        } = expenseData;

        const expense = { description, note, amount, createdAt };
        
        // add expense to firebase
        return database.ref(`users/${uid}/expenses`).push(expense).then((ref) => {
            dispatch(addExpense({
                id: ref.key,
                ...expense
            }));
        }).catch((e) => {
            console.log('startAddExpense Error:', e);
        });
    };
};

////////// REMOVE EXPENSE //////////

export const removeExpense = ({ id } = {}) => ({
    type: 'REMOVE_EXPENSE',
    id
});

export const startRemoveExpense = ({ id } = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/expenses/${id}`)
            .remove()
            .then(() => {
                dispatch(removeExpense({ id }));
            }).catch((e) => {
                console.log('startRemoveExpense Error:', e);
            });
    };
};

////////// EDIT EXPENSE //////////

export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
});

export const startEditExpense = (id, updates) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/expenses/${id}`)
            .update(updates)
            .then(() => {
                dispatch(editExpense(id, updates));
            }).catch((e) => {
                console.log('startEditExpense Error:', e);
            });
    };
};

////////// SET EXPENSES //////////

export const setExpenses = (expenses) => ({
    type: 'SET_EXPENSES',
    expenses
});

export const startSetExpenses = () => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;

        // get all expenses from firebase
        return database.ref(`users/${uid}/expenses`)
            .once('value')
            .then((snapshot) => {
                const expenses = [];
                snapshot.forEach((childSnapshot) => {
                    expenses.push({
                        id: childSnapshot.key,
                        ...childSnapshot.val()
                    });
                });

                dispatch(setExpenses(expenses));
            }).catch((e) => {
                console.log('startSetExpenses Error:', e);
            });
    };
};