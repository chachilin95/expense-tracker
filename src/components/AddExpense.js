import React from 'react';

export default class AddExpense extends React.Component {
    render() {
        return (
            <div>
                This is the Add Expense Page
            </div>
        );
    }
}