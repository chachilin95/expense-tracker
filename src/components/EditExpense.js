import React from 'react';

export default class EditExpense extends React.Component {
    render() {
        return (
            <div>
                This is the Edit Expense Page
            </div>
        );
    }
}