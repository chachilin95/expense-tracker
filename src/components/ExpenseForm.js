import React from 'react';

export default class ExpenseForm extends React.Component {
    state = {
        description: '',
        amount: 0.00,
        note: ''
    };

    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => ({ description }));
    };

    onAmountChange = (e) => {
        const dollarFormat = /^\d+(\.\d{0,2})?$/;
        const amount = e.target.value;

        if (amount.match(dollarFormat)) {
            this.setState(() => ({ amount }));
        }
    };

    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(() => ({ note }));
    };

    render() {
        return (
            <div>
                <form>
                    <input 
                        type='text'
                        placeholder='Description'
                        autoFocus
                        value={this.state.description}
                        onChange={this.onDescriptionChange}/>
                    <input 
                        type='number'
                        placeholder='Amount'
                        value={this.state.amount}
                        onChange={this.onAmountChange}/>
                    <textarea
                        placeholder='Add a note... (optional)'
                        value={this.state.note}
                        onChange={this.onNoteChange}/>
                    <button>Add Expense</button>
                </form>
            </div>
        );
    }
}