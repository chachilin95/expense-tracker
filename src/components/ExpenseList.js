import React from 'react';
import { connect } from 'react-redux';

const ExpenseList = (props) => (
    <div>
        <h1>Expense List</h1>
        <p>filters: {props.filters.text}</p>
        <p>expenses: {props.expenses.length}</p>
    </div>
);

const mapStateToProps = (state) => {
    return {
        expenses: state.expenses,
        filters: state.filters
    };
};

export default connect(mapStateToProps)(ExpenseList);