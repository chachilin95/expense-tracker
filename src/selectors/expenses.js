export default (expenses, { text, sortBy, startDate, endDate }) => {
    return expenses.filter((expense) => {
        const startDateMatch = typeof startDate !== 'number' || expense.createdAt >= startDate;
        const endDateMatch = typeof endDate !== 'number' || expense.createdAt <= endDate;
        const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());

        return startDateMatch && endDateMatch && textMatch;
    }).sort((e1, e2) => {
        if (sortBy === 'date') {
            return e1.createdAt < e2.createdAt ? 1 : -1;
        } else if (sortBy === 'amount') {
            return e1.amount < e2.amount ? 1 : -1;
        }
    });
};