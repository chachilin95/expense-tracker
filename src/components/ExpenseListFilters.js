import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import uuid from 'uuid';
import { setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate } from '../actions/filters';

export class ExpenseListFilters extends React.Component {
    state = {
        calendarFocused: null,
        startDateId: uuid(),
        endDateId: uuid()
    };

    onDatesChange = ({ startDate, endDate }) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    };

    onFocusChange = (calendarFocused) => {
        this.setState(() => ({ calendarFocused }));
    };

    onTextChange = (e) => {
        this.props.setTextFilter(e.target.value);
    };

    onSortChange = (e) => {
        if (e.target.value === 'date') {
            this.props.sortByDate();
        } else if (e.target.value === 'amount') {
            this.props.sortByAmount();
        }
    };

    render() {
        return (
            <div className='content-container'>
                <div className='input-group'>
                    <div className='input-group__item'>
                        <input
                            className='text-input'
                            onChange={this.onTextChange}
                            placeholder='Search expenses'
                            type='text'
                            value={this.props.filters.text}/>
                    </div>
                    <div className='input-group__item'>
                        <select
                            className='select'
                            onChange={this.onSortChange}
                            value={this.props.filters.sortBy}>
                            <option value='date'>Date</option>
                            <option value='amount'>Amount</option>
                        </select>
                    </div>
                    <div className='input-group__item'>
                        <DateRangePicker
                            startDate={this.props.filters.startDate}
                            startDateId={this.state.startDateId}
                            endDate={this.props.filters.endDate}
                            endDateId={this.state.endDateId}
                            onDatesChange={this.onDatesChange}
                            focusedInput={this.state.calendarFocused}
                            onFocusChange={this.onFocusChange}
                            numberOfMonths={1}
                            isOutsideRange={() => false}
                            showClearDates />
                    </div>
                </div>
            </div>
        );
    }
}

const mapStateToProps = (state) => ({
    filters: state.filters
});

const mapDispatchToProps = (dispatch) => ({
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    sortByDate: () => dispatch(sortByDate()),
    sortByAmount: () => dispatch(sortByAmount()),
    setStartDate: (date) => dispatch(setStartDate(date)),
    setEndDate: (date) => dispatch(setEndDate(date))
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);