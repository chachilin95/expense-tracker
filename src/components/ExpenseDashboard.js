import React from 'react';
import ExpenseList from './ExpenseList';

export default class ExpenseDashboard extends React.Component {
    render() {
        return (
            <div>
                <ExpenseList/>
            </div>
        );
    }
}