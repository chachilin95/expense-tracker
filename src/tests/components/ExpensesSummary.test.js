import React from 'react';
import { shallow } from 'enzyme';
import { ExpensesSummary } from '../../components/ExpensesSummary';
import expenses from '../fixtures/expenses';

test('Should render ExpensesSummary with no expenses', () => {
    const wrapper = shallow(<ExpensesSummary expenseCount={0} expensesTotal={0} />);
    expect(wrapper).toMatchSnapshot();
});

test('Should render ExpensesSummary with one expense', () => {
    const expense = expenses[1];
    const wrapper = shallow(<ExpensesSummary expenseCount={1} expensesTotal={expense.amount} />);
    expect(wrapper).toMatchSnapshot();
});

test('Should render ExpensesSummary with multiple expenses', () => {
    const total = expenses.reduce((sum, expense) => sum + expense.amount, 0);
    const wrapper = shallow(<ExpensesSummary expenseCount={expenses.length} expensesTotal={total} />);
    expect(wrapper).toMatchSnapshot();
});