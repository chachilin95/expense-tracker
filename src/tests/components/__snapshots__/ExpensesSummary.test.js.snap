// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render ExpensesSummary with multiple expenses 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        3
      </span>
       
      expenses
       totalling 
      <span>
        $1,275.00
      </span>
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;

exports[`Should render ExpensesSummary with no expenses 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        0
      </span>
       
      expenses
       totalling 
      <span>
        $0.00
      </span>
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;

exports[`Should render ExpensesSummary with one expense 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        1
      </span>
       
      expense
       totalling 
      <span>
        $95.00
      </span>
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;
